### PR TITLE
Created tests and the 'valid_move?' method for the Rook piece

### PR DESCRIPTION
**Pull request by Linda091**
Create the valid_move? logic for rook pieces, but my vertical tests kept failing. 

On line 9 in the models/rook.rb file, I initially had **return !v_obs?(x,y)**.

That piece of code was supposed to return **true** because _v_obs?(x,y)_ should've outputted **false**. 

However, I kept receiving an error for that line, even though the logic makes sense. I removed the ! and now it works, but the logic doens't make sense because by removing !, it returns true as in "yes, it is vertically obstructed". 

I could be missing something, but any insight would be appreciated. Thanks!

screen shot 2018-05-31 at 10 34 17 am

### CODE REVIEW

File: spec/models/rook_spec.rb [hunk 1]
```diff
@@ -1,4 +1,39 @@
...
+  it "shouldn't allow moves outside the chessboard" do
+    game = FactoryBot.create(:game)
+    piece1 = Rook.create(location_x: 0, location_y: 0, game_id: game.id)
+    expect(piece1.valid_move?(8,8)).to be false
+
+  end
+
 end
+
+=begin
```
Comment (Linda091) [inline spec/models/rook_spec.rb]: These were notes for my mentor. Forgot to remove them but will after the code is reviewed.

Comment (msarit) [reply spec/models/rook_spec.rb]: Hi Linda, I copied your code and played with it a bit.

Here's **rook.rb**:

```
class Rook < Piece

  def valid_move?(x,y)
    return false unless super
    return false unless valid_path?(x,y)
    return false if v_obs?(x,y) || h_obs?(x,y)
    return true
  end

  # This method checks whether the Rook travels a valid path
  def valid_path?(x,y)
    return true if y == self.location_y && x != self.location_x # horizontal move
    return true if y != self.location_y && x == self.location_x # vertical move
    return false # neither horizontal nor vertical move
  end
end
```

Here is **rook_spec.rb** _(RSpec tests; they all passed)_:
```
require 'rails_helper'

RSpec.describe Rook, type: :model do
  # when the game is created, it generates pieces, which may block the test lol

  it "should allow a rook to move vertically on board" do
    game = FactoryBot.create(:game)
    whiteRook = Rook.create(game_id: game.id, location_x: 0, location_y: 3)
    
    move = whiteRook.valid_move?(0,5)
    expect(move).to be true
  end

  it "should allow a rook to move horizontally on board" do
    game = FactoryBot.create(:game)
    whiteRook = Rook.create(game_id: game.id, location_x: 0, location_y: 3)
    
    move = whiteRook.valid_move?(3,3)
    expect(move).to be true
  end

  it "shouldn't allow moves outside the chessboard" do
    game = FactoryBot.create(:game)
    whiteRook = Rook.create(game_id: game.id, location_x: 0, location_y: 3)
    
    move = whiteRook.valid_move?(8,3)
    expect(move).to be false
  end

  it "shouldn't allow moves past an obstruction" do
    game = FactoryBot.create(:game)
    whiteRook = Rook.create(game_id: game.id, location_x: 0, location_y: 3)
    pawn = Pawn.create(game_id: game.id, location_x: 2, location_y: 3)
    
    move = whiteRook.valid_move?(3,3)
    expect(move).to be false
  end
end
```

I think the main thing with the tests was realizing that creating a game also creates pieces 😄 So you wanna make sure that the test_pieces you're creating are not clashing with the game's pieces (i.e. create test_pieces closer to the center of the board).

I also added a test to check that Rook can't move past an obstruction 👍

File: app/models/rook.rb [hunk 2]
```diff
@@ -1,2 +1,22 @@
 class Rook < Piece
+
+  def valid_move?(x,y)
+    return false unless super(x,y)              #if move is invalid, it will return false immediately vs going thrrough code line by line - saves time. 
+    current_loc_x = self.location_x             
+    current_loc_y = self.location_y 
+    
+    if current_loc_x == x && current_loc_y != y #checks if move is valid vertically
+      return v_obs?(x,y)                        # i initially had "return !v_obs?(x,y)" because it's supposed to return opp of v_obs. so if v_obs is false, returns true,
```
Comment (ilyakrasnov) [inline app/models/rook.rb]: I think it should be `!v_obs(x,y)` @Linda091

File: app/models/rook.rb [hunk 3]
```diff
@@ -1,2 +1,22 @@
...
+    current_loc_y = self.location_y 
+    
+    if current_loc_x == x && current_loc_y != y #checks if move is valid vertically
+      return v_obs?(x,y)                        # i initially had "return !v_obs?(x,y)" because it's supposed to return opp of v_obs. so if v_obs is false, returns true, 
+    end                                         # i.e. it is true the move is valid because it's not obstructed, but for some reason that kept causing my test to fail. i removed the ! and now it works but the logic doesn't make sense because return v_obs?(x,y) is basically "return false" because it's not obstructed and I want it to return true as in "it's true, it's valid"
+
+    if current_loc_y == y && current_loc_x != x #checks if move is valid horizontally
+      return !h_obs?(x,y)                       # returns opp of h_obs. so if h_obs is false, returns true, 
+    end                                         # i.e. it is true the move is valid
+
```
Comment (ilyakrasnov) [inline app/models/rook.rb]: Seems solid .. we can do the if's in 2 short lines here: 
```
return !v_obs?(x,y)  if current_loc_x == x && current_loc_y != y
return !h_obs?(x,y)  if current_loc_y == y && current_loc_x != x

```

File: spec/models/rook_spec.rb (hunk 1 above)
Comment (ilyakrasnov) [reply spec/models/rook_spec.rb]: Good catch with the pieces that get created @msarit

File: app/models/rook.rb (hunk 3 above)
Comment (Linda091) [reply app/models/rook.rb]: Agree. WiFi is spotty so don’t think I can edit my code while I’m here. Will edit when I get back.

File: app/models/rook.rb (hunk 2 above)
Comment (Linda091) [reply app/models/rook.rb]: @ilyakrasnov I agree, but that line of code still doesn't work, even though it should. I made sure my test_pieces started closer to the center, but I still receive the same vertical move error.

File: spec/models/rook_spec.rb (hunk 1 above)
Comment (Linda091) [reply spec/models/rook_spec.rb]: @msarit good point w/ the pieces that get created. i think that was the reason why my code wasn't working. edited it and it works now :)

File: app/models/rook.rb (hunk 2 above)
Comment (Linda091) [reply app/models/rook.rb]: nevermind - the code works now :)

File: spec/models/rook_spec.rb [hunk 4]
```diff
@@ -1,4 +1,30 @@
...
+    rook = Rook.create(location_x: 5, location_y: 5, game_id: game.id)
+    expect(rook.valid_move?(8,8)).to be false
+  end
+
+  it "shouldn't allow rook to move past an obstruction" do
+    game = FactoryBot.create(:game)
+    rook = Rook.create(location_x: 0, location_y: 3, game_id: game.id)
+    pawn = Pawn.create(location_x: 0, location_y: 5, game_id: game.id)
+    expect(rook.valid_move?(0,7)).to be false
+  end
```
Comment (ilyakrasnov) [inline spec/models/rook_spec.rb]: lets add add a few test cases for non-linear move - just to make sure 😄

**Comment by ilyakrasnov** [comment]
Looks solid @Linda091, great work! I added a small comment on the tests - once that is addressed, there are some merge conflicts - so you can't merge the branch directly without resolving them... 

Best procedure would be:

* pull current master on your machine
* switch to your `create-valid-moves-for-rook` branch
* run `git merge master` - you will see conflicts in the 3 files above
* resolve the conflicts locally and commit
* push the `create-valid-moves-for-rook` branch to github (this will update the pr and now there will be no conflicts)
* merge the pr as usual

File: spec/models/rook_spec.rb (hunk 4 above)
Comment (Linda091) [reply spec/models/rook_spec.rb]: @ilyakrasnov meaning diagonal moves? If so, I added the last test case w/ a few diagonal examples.
screen shot 2018-06-07 at 11 44 53 am

**Pull request merged**